File: is_int/c#/is_int_flag.cs
using System;

class is_int_flag {
	static bool IsInt(string s) {
		var state = 0;
		foreach (byte b in System.Text.Encoding.UTF8.GetBytes(s)) {
			if (state == 0) {
				if (b == '+' || b == '-') {
					state = 1;
				} else if ('0' <= b && b <= '9') {
					state = 2;
				} else {
					state = -1;
				}
			} else if (state == 1) {
				if ('0' <= b && b <= '9') {
					state = 2;
				} else {
					state = -1;
				}
			} else if (state == 2) {
				if ('0' <= b && b <= '9') {
					state = 2;
				} else {
					state = -1;
				}
			}
			if (state < 0) {
				break;
			}
		}
		return state == 2;
	}

	static void Main() {
		(string s, bool b)[] tests = {
			("", false),
			("+", false),
			("-", false),
			("++", false),
			("--", false),
			("+0+", false),
			("-0-", false),
			("+0123456789", true),
			("-0123456789", true),
			("0123456789", true),
		};

		foreach (var t in tests) {
			Console.WriteLine("'{0}'\t{1}", t.s, t.b == IsInt(t.s) ? "OK" : "NG");
		}

	}
}

File: is_int/c/is_int_flag.c
#include <stdio.h>

int next_state[][4] = {
	{-2,  1,  2, -2},
	{-2, -2,  2, -2},
	{-2, -2,  2, -1},
};

int char_kind(char c)
{
	if (c == '+' || c == '-') {
		return 1;
	} else if ('0' <= c && c <= '9') {
		return 2;
	} else if (c == '\0') {
		return 3;
	} else {
		return 0;
	}
}

int is_int(char *s)
{
	int state = 0;
	while (state >= 0) {
		state = next_state[state][char_kind(*s)];
		s++;
	}

	return state == -1;
}

int main(void)
{
	struct {
		char *s;
		int b;
	} tests[] = {
		{ "", 0 },
		{ " ", 0 },
		{ "+", 0 },
		{ "-", 0 },
		{ "++", 0 },
		{ "--", 0 },
		{ "+0+", 0 },
		{ "-0-", 0 },
		{ "+0123456789", 1 },
		{ "-0123456789", 1 },
		{ "0123456789", 1 }
	};
	int i;

	for (i = 0; i < sizeof(tests)/sizeof(*tests); i++) {
		printf("'%s'\t%s\n", tests[i].s, tests[i].b == is_int(tests[i].s) ? "OK" : "NG"); 
	}

	return 0;
}
